Pull request for IBM Power Support

drchristensen opened the pull request:
I've created a patch series which adds support for IBM Power systems.  These changes should be compatible with x86_64 systems but has only cursory testing on the x86_64 platform (i.e. running get_cpulist.sh to verify output is as expected).  Some of the changes include:

- Add support for a custom yum repo (seperate from existing subscription and QE magic)
- Moved several hard-coded settings from pvp_ovsdpdk.yml to test-settings.yml to allow supporting the Power platform more easily (VM CPU isolation list, number of driver queues, IOMMU settings, etc.)
- Temporary workarounds for issues with upstream software (use of uio_pci_generic in the VM, OVS requirement for sequential NUMA numbering)

Individual patches are useful to code review to understand the changes involved but the entire patch series should be applied together to ensure they work correctly.

drchristensen committed:
Reorganize test_settings layout

Reorganized test_settings file and organized settings by target (DUT,
T-Rex server, VM). Shortened some descriptions to less than 80
characters.  Added missing vm_driver setting.

Signed-off-by: David Christensen <drc@ibm.com>

drchristensen committed:
Added missing vm_interface_X_pciid settings

Signed-off-by: David Christensen <drc@ibm.com>

drchristensen committed:
Fix issue with custom repo mode setting

Fixed inconsistent naming for custom_repo_mode setting.

Signed-off-by: David Christensen <drc@ibm.com>

drchristensen committed:
Add support for custom yum repos

Add a step to the Ansible script to setup a custom yum repo, add
variables to test_settings.yml to select and access the repo.

Signed-off-by: David Christensen <drc@linux.vnet.ibm.com>

drchristensen committed:
Enable more DUT kernel command-line and tuned tuning options

Add additional settings for DUT and T-Rex to allow architecture
specific kernel command line tuning.  Modify Ansible script logic
to change existing options if present, add options if not present,
and only reboot the system if any options have changed.  Change
logic for enabling cpu-partitioning profile for tuned.

Signed-off-by: David Christensen <drc@linux.vnet.ibm.com>

drchristensen committed:
Add arch specific CPU support, tuning script

Change settings so that only the PCI address of the NICs are used.
Substantial changes to get_cpulist script to support Power architecture,
add more v_cpus, generate number of driver RX queues, NUMA node for
VM memory tuning. Add script to tune the PCIe MaxReadReq settings for
Power systems.

Signed-off-by: David Christensen <drc@linux.vnet.ibm.com>

drchristensen committed:
Set OVS/QEMU permissions for Power, cleanup some play names

Add Ansible plays to configure OVS & QEMU to run on Power.  Rename
some plays to be more consistent.

Signed-off-by: David Christensen <drc@linux.vnet.ibm.com>

drchristensen committed:
Modify VM setup and configuration to run on Power

Several tweaks to the VM configuration to run on Power.  Allows
uio_pci_generic to be used instead of vfio-pci which is not currently
supported in a VM.

Signed-off-by: David Christensen <drc@linux.vnet.ibm.com>

ctrautma commented on ansible/test_settings.yml:
```diff
@@ -1,37 +1,62 @@
+# server.
 trex_path: https://trex-tgn.cisco.com/trex/release
-trex_version: v2.53
+trex_version: v2.72
```
Need to check with Eelco if his scripts are supporting version 2.72.  I believe at one point updating past this version was an issue.  Will double check.

chaudron replied:
The update for the newer API was added previously, so if the test works the upgrade should be fine. It would be interesting to do the actual run with both versions to see if there is any performance degradation.

ctrautma commented on ansible/get_cpulist.sh:
```diff
@@ -33,100 +73,207 @@ then
-first_ht_pair=$(cat /sys/devices/system/cpu/cpu${dev_numa_node}/topology/thread_siblings_list | tr , " ")
-dut_isolated_cpus=$numa_cpulist
+# Make sure the system has the minimum required NUMA resources
+req_numa_count=2
```
We can run this on single numa systems.  This was an issue when we first did a lot of work on these tests as the lab system was single numa.  We should support this.  Anyway to add something here for single numa?

ctrautma commented on ansible/pvp_ovsdpdk.yml:
```diff
@@ -349,13 +433,13 @@
 
   - name: Setting up OVS-DPDK config step 1
-    shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-init=true"
+    shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-socket-mem=4096,0,0,0,0,0,0,0,4096"
```
Will this work with any numa config?  Or should this be a configuration option?

ctrautma reviewed (commented):
Hi David,

Thanks for this patch,  just a few commnets/questions when you get a moment.

drchristensen commented:
Regarding trex_version, no problem staying with the older version, I picked a newer release because it was available.  I'm not seeing any issues with the testing, but it's my first time running the tests so I don't have a reference point to know if something is wrong.
 
 
The requirement for 2 NUMA nodes isn't a Power requirement, it was present in get_cpulist.sh before my changes (see snippets below).  There shouldn't be a problem supporting everything on a single NUMA node, just need changes for both x86 and Power to reserve the necessary CPUs.
 
  63 nu_numa_node=$(lscpu | grep "NUMA node(s)" | awk '{print $3}')  64 # need at least two numa node  65 if [ $nu_numa_node -lt 2 ]  66 then  67   exit 1  68 fi
 
and 
 
 125 dut_dpdk_lcore_mask() 126 { 127 | cpu_on_anothernuma=$(lscpu | grep "NUMA node.*CPU" | grep -v "NUMA node${dev_numa_node}" | head -n 1 | awk '{print $4}') 128 | cpu_dpdk_lcore=$(echo $cpu_on_anothernuma | awk -F, '{print $2}') 129 | pmd_mask "$cpu_dpdk_lcore" 130 }
 
 
Regarding the ovs-vsctl change, there's a limitation in the current OVS design that expects NUMA nodes to be sequentially numbered (0, 1, 2, ...).  That assumption is not correct for Power systems:
 
$ lscpuArchitecture:        ppc64leByte Order:          Little EndianCPU(s):              128On-line CPU(s) list: 0-127Thread(s) per core:  4Core(s) per socket:  16Socket(s):           2NUMA node(s):        6Model:               2.3 (pvr 004e 1203)Model name:          POWER9, altivec supportedCPU max MHz:         3800.0000CPU min MHz:         2300.0000L1d cache:           32KL1i cache:           32KL2 cache:            512KL3 cache:            10240KNUMA node0 CPU(s):   0-63NUMA node8 CPU(s):   64-127NUMA node252 CPU(s):NUMA node253 CPU(s):NUMA node254 CPU(s):NUMA node255 CPU(s):
 
This limitation really only affects the default DPDK EAL --socket-mem initialization value, so the local workaround requires a customized openvswitch RPM and the ovs-vsctl command you see. We're in the process of submitting patches to the upstream OVS project to remove this limitation, so the Ansible current code is more of a hack to workaround this issue.  I'm open to ideas on how to manage this problem until we have code accepted upstream and an updated openvswitch RPM is released.  
 
Dave
 
----- Original message -----From: Christian Trautman <notifications@github.com>To: ctrautma/RHEL_NIC_QUALIFICATION <RHEL_NIC_QUALIFICATION@noreply.github.com>Cc: David Christensen <drc@ibm.com>, Author <author@noreply.github.com>Subject: [EXTERNAL] Re: [ctrautma/RHEL_NIC_QUALIFICATION] Pull request for IBM Power Support (#23)Date: Wed, Feb 5, 2020 9:54 AM 
@ctrautma commented on this pull request.
Hi David,
Thanks for this patch, just a few commnets/questions when you get a moment.
In ansible/test_settings.yml:
>  trex_path: https://trex-tgn.cisco.com/trex/release-trex_version: v2.53+trex_version: v2.72
Need to check with Eelco if his scripts are supporting version 2.72. I believe at one point updating past this version was an issue. Will double check.
In ansible/get_cpulist.sh:
> @@ -33,100 +73,207 @@ then  exit 1 fi -dev_numa_node=$(lspci  -s $dev_pci -v | grep -o "NUMA node [0-9]*" | awk '{print $3}')-numa_cpulist=$(lscpu | grep "NUMA node${dev_numa_node}" | awk '{print $4}')-first_ht_pair=$(cat /sys/devices/system/cpu/cpu${dev_numa_node}/topology/thread_siblings_list | tr , " ")-dut_isolated_cpus=$numa_cpulist+# Make sure the system has the minimum required NUMA resources+req_numa_count=2
We can run this on single numa systems. This was an issue when we first did a lot of work on these tests as the lab system was single numa. We should support this. Anyway to add something here for single numa?
In ansible/pvp_ovsdpdk.yml:
> @@ -349,13 +433,13 @@ ############################    - name: Setting up OVS-DPDK config step 1-    shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-init=true"+    shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-socket-mem=4096,0,0,0,0,0,0,0,4096"
Will this work with any numa config? Or should this be a configuration option?
—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
 

ctrautma commented:
OK let me just get feedback from Eelco on the version for the T-Rex item. 

I will ask the get_cpulist.sh author to accommodate single numa configs. As far as how to handle the Power issue, I am unsure....maybe there is someway to use the get_cpulist.sh script to parse the cpu info to calculate a correct memory config that can be passed onto the scripts. 

-Christian, 


----- Original Message -----

From: "David Christensen" <notifications@github.com> 
To: "ctrautma/RHEL_NIC_QUALIFICATION" <RHEL_NIC_QUALIFICATION@noreply.github.com> 
Cc: "Christian Trautman" <ctrautma@redhat.com>, "Mention" <mention@noreply.github.com> 
Sent: Wednesday, February 5, 2020 2:29:51 PM 
Subject: Re: [ctrautma/RHEL_NIC_QUALIFICATION] Pull request for IBM Power Support (#23) 

Regarding trex_version, no problem staying with the older version, I picked a newer release because it was available. I'm not seeing any issues with the testing, but it's my first time running the tests so I don't have a reference point to know if something is wrong. 


The requirement for 2 NUMA nodes isn't a Power requirement, it was present in get_cpulist.sh before my changes (see snippets below). There shouldn't be a problem supporting everything on a single NUMA node, just need changes for both x86 and Power to reserve the necessary CPUs. 

63 nu_numa_node=$(lscpu | grep "NUMA node(s)" | awk '{print $3}') 64 # need at least two numa node 65 if [ $nu_numa_node -lt 2 ] 66 then 67 exit 1 68 fi 

and 

125 dut_dpdk_lcore_mask() 126 { 127 | cpu_on_anothernuma=$(lscpu | grep "NUMA node.*CPU" | grep -v "NUMA node${dev_numa_node}" | head -n 1 | awk '{print $4}') 128 | cpu_dpdk_lcore=$(echo $cpu_on_anothernuma | awk -F, '{print $2}') 129 | pmd_mask "$cpu_dpdk_lcore" 130 } 


Regarding the ovs-vsctl change, there's a limitation in the current OVS design that expects NUMA nodes to be sequentially numbered (0, 1, 2, ...). That assumption is not correct for Power systems: 

$ lscpuArchitecture: ppc64leByte Order: Little EndianCPU(s): 128On-line CPU(s) list: 0-127Thread(s) per core: 4Core(s) per socket: 16Socket(s): 2NUMA node(s): 6Model: 2.3 (pvr 004e 1203)Model name: POWER9, altivec supportedCPU max MHz: 3800.0000CPU min MHz: 2300.0000L1d cache: 32KL1i cache: 32KL2 cache: 512KL3 cache: 10240KNUMA node0 CPU(s): 0-63NUMA node8 CPU(s): 64-127NUMA node252 CPU(s):NUMA node253 CPU(s):NUMA node254 CPU(s):NUMA node255 CPU(s): 

This limitation really only affects the default DPDK EAL --socket-mem initialization value, so the local workaround requires a customized openvswitch RPM and the ovs-vsctl command you see. We're in the process of submitting patches to the upstream OVS project to remove this limitation, so the Ansible current code is more of a hack to workaround this issue. I'm open to ideas on how to manage this problem until we have code accepted upstream and an updated openvswitch RPM is released. 

Dave 

----- Original message -----From: Christian Trautman <notifications@github.com>To: ctrautma/RHEL_NIC_QUALIFICATION <RHEL_NIC_QUALIFICATION@noreply.github.com>Cc: David Christensen <drc@ibm.com>, Author <author@noreply.github.com>Subject: [EXTERNAL] Re: [ctrautma/RHEL_NIC_QUALIFICATION] Pull request for IBM Power Support (#23)Date: Wed, Feb 5, 2020 9:54 AM 
@ctrautma commented on this pull request. 
Hi David, 
Thanks for this patch, just a few commnets/questions when you get a moment. 
In ansible/test_settings.yml: 
> trex_path: https://trex-tgn.cisco.com/trex/release-trex_version: v2.53+trex_version: v2.72 
Need to check with Eelco if his scripts are supporting version 2.72. I believe at one point updating past this version was an issue. Will double check. 
In ansible/get_cpulist.sh: 
> @@ -33,100 +73,207 @@ then exit 1 fi -dev_numa_node=$(lspci -s $dev_pci -v | grep -o "NUMA node [0-9]*" | awk '{print $3}')-numa_cpulist=$(lscpu | grep "NUMA node${dev_numa_node}" | awk '{print $4}')-first_ht_pair=$(cat /sys/devices/system/cpu/cpu${dev_numa_node}/topology/thread_siblings_list | tr , " ")-dut_isolated_cpus=$numa_cpulist+# Make sure the system has the minimum required NUMA resources+req_numa_count=2 
We can run this on single numa systems. This was an issue when we first did a lot of work on these tests as the lab system was single numa. We should support this. Anyway to add something here for single numa? 
In ansible/pvp_ovsdpdk.yml: 
> @@ -349,13 +433,13 @@ ############################ - name: Setting up OVS-DPDK config step 1- shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-init=true"+ shell: "ovs-vsctl set Open_vSwitch . other_config:dpdk-socket-mem=4096,0,0,0,0,0,0,0,4096" 
Will this work with any numa config? Or should this be a configuration option? 
—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe. 


— 
You are receiving this because you were mentioned. 
Reply to this email directly, view it on GitHub , or unsubscribe .

drchristensen committed:
Fixed a problem when initializing OVS/DPDK on x86_64 systems

Current playbook includes a command to set the DPDK "socket-mem"
parameter when starting OVS:

dpdk-socket-mem=4096,0,0,0,0,0,0,0,4096

This setting is only effective for Power systems due to an OVS
limitation when handling non-consecutive NUMA nodes.  Make the
current command applicable to Power architecture only and add a
new command that appropriate for x86_64 systems:

dpdk-socket-mem=4096

Signed-off-by: David Christensen <drc@ibm.com>

ctrautma commented:
Hi David,

just following up on this as I am ready to merge this once we come to a consensus on this issue.
"""
Regarding the ovs-vsctl change, there's a limitation in the current OVS design that expects NUMA nodes to be sequentially numbered (0, 1, 2, ...). That assumption is not correct for Power systems: 
"""

Is making this a configurable option in the yaml file a possibility?  I just worry with our usual two numa config in x86 that this will break that,  but I have not tested it.  I hope I am not reading your change incorrectly so please correct me if I am wrong here.  We usually use 4096, 4096 in x_86 configs in our lab.
